# CS-GY-66143 - MIDTERM EXAM SPRING 2024

Please upload this problem in Brightspace together with your answers to the in-person portion. You need to submit your github URL and include this notebook with your answers for 1A and 1C in the same folder that you have answered 1B.  

## PROBLEM 1: A JOURNEY THROUGH INFORMATION THEORY AND NEURAL NETWORKS (30 POINTS)

### Introduction

Information theory was introduced by Claude Shannon in 1948. It is a mathematical theory that deals with the transmission, processing, utilization, and extraction of information. It has given rise to a wide range of applications, including data compression, cryptography, error correction and fueled other industries such as AI, cellular communications and others.

In this problem, we will use key concepts from information theory aiming in opening up the internals of neural networks and potentially explain how statistical learning theory and information theory can explain their behavior.

The journey begins with [this reference](https://arxiv.org/abs/2206.07867) that you need to study before attempting the problem below.

### PS 1A: Information Theory Basics (5 points)

Let (x, y) have the following joint distribution:

![](joint-distribution.png)

If H is the symbol for the entropy functional, answer quantitatively showing your calculations

a.  if H(x|y)=H(y|x) and if H(x) - H(x|y) =H(y)-H(y|x).

b. Calculate the mutual information I(x,y)



# Solution Part A:

$$H(x \mid y) = −\sum_{i, j}​P(xi​,yj​)⋅\log_2​(P(x_i \mid y_j​)$$

and

$$ P(x|y) = \frac{P(x, y)}{P(y)} $$

and

$$H(y \mid x) = −\sum_{i, j}​P(x_i​,y_j​)⋅\log_2​(P(y_j​,x_i​)P(x_i​)$$

and

$$H(x) = −\sum_{i}​P(x_i​​)⋅\log_2​P(x_i​)$$

In [7]:
from math import log2

def calculate_conditional_entropy(joint_probs, marginal_probs):
    conditional_entropy = 0

    for y in range(len(joint_probs[0])):
        for x in range(len(joint_probs)):
            joint_prob = joint_probs[x][y]
            if joint_prob > 0:  # Exclude cases where joint probability is zero
                conditional_prob = joint_prob / marginal_probs[y]
                conditional_entropy += joint_prob * log2(conditional_prob)

    return -conditional_entropy

def calculate_entropy(marginal_probs):
    entropy = 0
    for prob in marginal_probs:
        if prob > 0:  # Exclude cases where probability is zero
            entropy += prob * log2(prob)
    return -entropy

joint_probs = [ # copied from the given joint distribution table
    [1/8, 1/16, 1/32, 1/32],
    [1/16, 1/8, 1/32, 1/32],
    [1/16, 1/16, 1/16, 1/16],
    [1/4, 0, 0, 0]
]

marginal_probs_y = [1/4, 1/4, 1/4, 1/4] # manually calculated as the sum of each row
marginal_probs_x = [1/2, 1/4, 1/8, 1/8] # manually calculated as the sum of each column

conditional_entropy_given_y = calculate_conditional_entropy(joint_probs, marginal_probs_y)
conditional_entropy_given_x = calculate_conditional_entropy(joint_probs, marginal_probs_x)

entropy_y = calculate_entropy(marginal_probs_y)
entropy_x = calculate_entropy(marginal_probs_x)

print("H(x|y) =", conditional_entropy_given_y)
print("H(y|x) =", conditional_entropy_given_x)
print()
print("H(x) - H(x|y) = ", entropy_x - conditional_entropy_given_y)
print("H(y) - H(y|x) = ", entropy_y - conditional_entropy_given_x)

H(x|y) = 1.375
H(y|x) = 1.625

H(x) - H(x|y) =  0.375
H(y) - H(y|x) =  0.375


#Summary

Given the above calculations, we can conclude that $H(x \mid y)$ does not equal $H(y \mid x)$, but that $H(x) - H(x \mid y)$ and $H(y) - H(y \mid x)$ are indeed equivalent. This is not surprising given that both $H(x) - H(x \mid y)$ and $H(y) - H(y \mid x)$, as well as $H(X) + H(Y) − H(X, Y)$, are all different forms of defining mutual information, $I(x; y)$, as stated on the bottom of page 14 of [A Visual Introduction to Information Theory](https://arxiv.org/pdf/2206.07867.pdf)

# Solution Part B:

As mentioned in the Summary of part A,

$$I(x; y) = H(x) - H(x \mid y) = H(y) - H(y \mid x)$$

As we have already calculated $H(x) - H(x \mid y) = H(y) - H(y \mid x) = 0.375$, we know that:

 $$I(x; y) = 0.375$$

### PS 1B: Tishby's Information Bottleneck (20 points)

In this seminal [paper](https://arxiv.org/abs/1703.00810), Tishby et al. propose a new framework for understanding the learning dynamics of deep neural networks. They argue that the learning process can be understood as an information bottleneck. This is [very nice video summary](https://www.youtube.com/watch?v=bLqJHjXihK8) of the key findings.

Clone this repo https://github.com/pantelis/IDNNs and using the VSCode remote container extension, open the cloned repo in the tensorflow container. Please note the Dockerfile specification under the /docker folder - you may need to change the Dockerfile to use a CPU, for example, base image depending on your environment.

a. Run the code in the repo and plot the training process on the information plane such as in the video and the figure below.

![](information-plane.jpg)

b. Add a folder `midterm-take-home` in your existing assignments repo.  Design a CNN classification network that can classify [the CIFAR10 dataset](https://huggingface.co/datasets/cifar10)  and introduce the information measures that the IDNN repo has introduced to produce the information plane and gradient mean and variance figures.

Note that IDNN repo is based on the old version of tensorflow and you need to update the code to the latest version of tensorflow or pytorch. Both frameworks they need to be in their latest (but stable) version. Luckily a good chunk of supporting functions are framework independent.

The new code should be in script *.py files but a tutorial notebook must also exist that links to the functions of the scripts and using markdown cells explains what the code is trying to do. Do not hesitate to use latex formulas to explain the concepts. Your focus in this tutorial treatment is on computational aspects and not on the information theoretic bounds outlined at some point in the video.  


### PS 1C: Commentary (5 points)

Write a commentary on the findings of your own experiment (1B) highlighting the key findings in a similar way that Tishby et al. have done in their paper / video.
